virtual_rainforest/models/abiotic/radiation.py

```diff
@@ -0,0 +1,208 @@
+# this doesn't exist yet; optional scipy
+CLOUDY_TRANSMISSIVITY = 0.25
+"""Cloudy transmittivity :cite:p:`Linacre1968`"""
+TRANSMISSIVITY_COEFFICIENT = 0.50
+"""Angular coefficient of transmittivity :cite:p:`Linacre1968`"""
+FLUX_TO_ENERGY = 2.04
+"""From flux to energy conversion, umol/J :cite:p:`Meek1984`"""
+BOLZMAN_CONSTANT = 5.67e-8
+"""Stephan Bolzman constant W m-2 K-4"""
+SOIL_EMISSIVITY = 0.95
+"""Soil emissivity, default for tropical rainforest"""
+CANOPY_EMISSIVITY = 0.95
+"""Canopy emissivity, default for tropical rainforest"""
+BEER_REGRESSION = 2.67e-5
+"""Parameter in equation for atmospheric transmissivity based on regression of Beer’s
+radiation extinction function :cite:p:`Allen1996`"""
+CELSIUS_TO_KELVIN = 273.15
+"""Factor to convert temperature in Celsius to absolute temperature in Kelvin"""
+SECOND_TO_DAY = 86400
+"""Factor to convert between days and seconds."""
+
+
+class Radiation:
+    """Radiation balance."""
+
+    def __init__(self, elevation: NDArray[np.float32]) -> None:
+        """Initializes point-based radiation method."""
+
+        self.elevation = elevation
+        """Elevation above sea level, [m]"""
+
+        # populated later
+        self.ppfd = NDArray[np.float32]
+        """Daily top of canopy photosynthetic photon flux density, [mol m-2]"""
+        self.topofcanopy_radiation = NDArray[np.float32]
+        """Daily top of canopy downward shortwave radiation, [J m-2]"""
+        self.longwave_canopy = NDArray[np.float32]
+        """Daily longwave radiation from n individual canopy layers, [J m-2]"""
+        self.longwave_soil = NDArray[np.float32]
+        """Daily longwave radiation from soil, [J m-2]"""
+        self.netradiation_surface = NDArray[np.float32]
+        """Daily net shortwave radiation at the surface (= forest floor), [J m-2]"""
+
+    def calc_ppfd(
+        self,
+        shortwave_in: NDArray[np.float32],
+        sunshine_fraction: NDArray[np.float32] = np.array(1.0, dtype=np.float32),
+        albedo_vis: NDArray[np.float32] = np.array(0.03, dtype=np.float32),
+    ) -> NDArray[np.float32]:
+        """Calculates top of canopy photosynthetic photon flux density [mol m-2].
+
+        Args:
+            shortwave_in: daily downward shortwave radiation [J m-2]
+            sunshine_fraction: fraction of sunshine hours, between 0 (100% cloud cover)
+                and 1 (cloud free sky)
+            albedo_vis: visible light albedo, default = 0.03
+
+        Returns:
+            tau: atmospheric transmissivity, unitless
+
+        Reference: :cite:t:`Davis2017`
+        """
+        # calculate transmissivity (tau), unitless
+        tau_o = CLOUDY_TRANSMISSIVITY + TRANSMISSIVITY_COEFFICIENT * sunshine_fraction
+        tau = tau_o * (1.0 + BEER_REGRESSION * self.elevation)
+
+        # Calculate daily photosynth. photon flux density (ppfd_d), mol/m^2
+        self.ppfd = (1.0e-6) * FLUX_TO_ENERGY * (1.0 - albedo_vis) * tau * shortwave_in
+        return tau
```

Comment:
The pure version of the functional programming paradigm will tell you that the function outputs should depend exclusively on its inputs. Based on that, and something I've used in the past and has came handy for testing, **even constants should be passed as inputs**:

```python
def calc_something(a, b, sigma = STEFAN_BOLTZMAN_CONSTANT):
    ...
```

I think this is overkilling for really mathematical constants (eg. pi), but it might be useful for constants that might not be constants all the time (eg. is it `CLOUDY_TRANSMISSIVITY = 0.25` always? Or is it something that depends on the specific model you are implementing?

This might linked with something mentioned above about the albedo.

Reply:
There's a parallel here to my `pyrealm` package which has a _ton_ of these kind of empirical 'constants'.

There, I made the decision that if people would typically be using a consistent set of those values across an analysis, then packaging them into a (user editable) `dataclass` of constants that could be used consistently was easier than having every last constant as a named input. So functions have a `params` input that accepts a `PModelParams` object. 

As well as keeping keeping the function/method signatures simple, this also isolates a particular set of values in a (frozen) dataclass so it is always clear which values are being used. You can't - for example - accidentally update `CONSTANTS.STEFAN_BOLTZMAN_CONSTANT` half way through. You can of course use two different parameter sets inconsistently but that's on you 😄 

This idea though has not been exposed to review...

Reply:
Honestly, I slightly prefer @dalonsoa's approach of the two. I'm not sure when you'll want to define a whole set of constants, other than the one place where you define them with their "real" values, so it makes more sense to me to pass in a single constant value to a pure function (e.g. for testing purposes). I might be missing something.

Either way though, it seems like we are broadly in agreement and the details don't matter so much.

Reply:
Thank you all for the suggestions, although I must admit I got a bit confused by the end of the day ;-) I see the point in adding constants as arguments to the functions, and to set defaults if they are not true constants, but then I no longer have the CONSTANTS in capitals, which makes it less clear when looking at the code.
I suggest to go back to an earlier suggestion and make more use of scipy (for example for Stefan Boltzmann constant) and other (more variable) inputs can be set to default but changes in one of the config files in the future?

Reply:
Let's not rush to do this. @vgro - can you leave those constants as they are (using global values) for the moment.

Reply:
We can always revisit this and move to the purer version but we have a functioning approach for making these initial drafts and we are going to want a coherent approach across all modules. The reason I'm suggesting waiting and addressing this in a later PR (maybe across modules) is that  this project is going to have an awful lot of parameters:

* Some of those are going to be 'real' constants which _might_ be in `scipy` but might not, so having a single source that brings them all together is going to be cleaner. Notably, `scipy` only provides 'true' constants and doesn't cover Earth specific constants like the adiabatic lapse rate etc.

* Some functions are going to use physical equations from the literature that have complex empirical parameterisation. Those are not things that a user is likely to want to tinker with: 'Hmm, I wonder what happens to elephants if the temperature / water density relationship changes?'.

  I don't think these should be hard coded into functions because they might be revised but if you take a purely functional approach and have them all as arguments, but then the function signatures can get very long. That isn't inherently a problem, but it does make the API feel a lot more complex. Essentially there are "args that are real data variables" and "lots of args you will never touch".

  Those are some of the reasons I took the parameter set approach in `pyrealm` - the most obvious application there is the set of `calc_` functions [here](https://github.com/davidorme/pyrealm/blob/master/pyrealm/pmodel.py#L128).

* And finally there are `vr` empirical 'not-constants-at-all', which are going to be empirical parameters. Those are likely to be values that should be specified in the config and are expected to be changed frequently between simulations. I think *those* should be pure function arguments!

I'm definitely not saying that the approach in `pyrealm` is better (or even good or well implemented) but it is a solution to something like the unusually complex parameterisation we will have. I just think it is worth leaving things as they are until we've got a feel for how we want to consider this holistically.

Reply:
> Let's not rush to do this. @vgro - can you leave those constants as they are (using global values) for the moment.

Agreed. Maybe we have ventured a bit off topic. Constants aside, the main thing I think is turning some of the member variables into input arguments, as we were initially talking about.

Reply:
Just tagging in #129 - and we should maybe convert that issue to a discussion.

We now have a couple of whole system things where we want to synchronise approaches: constants and also using `pint` to add units to everything. Both of those should be adopted universally so it seems better to continue with the individual module developments and then have a joint branch/PR to adopt those overarching strategies once we've firmed up what we want to do.